File: Dynamic Programming/97. Interleaving String.py
class Solution:
    
    
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False 
        
        
        ptr1 = 0 
        ptr2 = 0 
        ptr3 = 0 
        cache = {}
        
        return self.recursive(s1, s2, s3, ptr1, ptr2, ptr3, cache)
        
    def recursive(self, s1, s2, s3, ptr1, ptr2, ptr3, cache):
        
        
        if ptr3 == len(s3) and ptr1 + ptr2 != ptr3:
            return False 
        
        elif ptr3 == len(s3) and ptr1 + ptr2 == ptr3:
            return True 
        
        elif ptr3 == len(s3):
            return False 
        
        if (ptr1, ptr2, ptr3) in cache:
            return cache[(ptr1, ptr2, ptr3)]
        
        cache[(ptr1, ptr2, ptr3)] = False 
        
        
        if ptr1 == len(s1) and ptr2 == len(s2):
            return False 
        
        if ptr1 < len(s1) and s3[ptr3] == s1[ptr1]:
            if self.recursive(s1, s2, s3, ptr1 + 1, ptr2, ptr3 + 1, cache):
                cache[(ptr1, ptr2, ptr3)] = True 
                return cache[(ptr1, ptr2, ptr3)]
            
        if ptr2 < len(s2) and s3[ptr3] == s2[ptr2]:
            if self.recursive(s1, s2,s3, ptr1, ptr2 + 1, ptr3 + 1, cache):
                cache[(ptr1, ptr2, ptr3)] = True 
                return cache[(ptr1, ptr2, ptr3)]
            
        
        return cache[(ptr1, ptr2, ptr3)] 
        

File: Dynamic Programming/97. Interleaving String.cpp
class Solution {
public:
    int n;
    int i, j;
    string goal;
    string str1;
    string str2;
    vector<vector<int>>dp;
    bool dfs(int ptr1, int ptr2) {
        
        if (ptr1 + ptr2 == n && ptr1 == i && ptr2 == j) return true;
        if (ptr1 >= i && ptr2 >= j) return false;
        if (dp[ptr1][ptr2] != -1) return dp[ptr1][ptr2];
        bool can = false;
        if (ptr1 < i && goal[ptr1 + ptr2] == str1[ptr1]) {
            can = can || dfs(ptr1 + 1, ptr2);
        }
        if (ptr2 < j && goal[ptr1 + ptr2] == str2[ptr2]) {
            can = can ||dfs(ptr1, ptr2 + 1);

        }
        dp[ptr1][ptr2] = can;
 
        return can;


    }
    bool isInterleave(string s1, string s2, string s3) {
        
        n = s3.length();
        i = s1.length();
        j = s2.length();
        dp = vector<vector<int>>(i + 1, vector<int>(j + 1, -1));
        goal = s3;
        str1 = s1, str2 = s2;
        if (i + j != n) return false;
        return dfs(0,0 );

    }
};